contracts/chain-adapters/Linea_Adapter.sol

```diff
@@ -0,0 +1,354 @@
+    error NotFromRemoteUSDCBridge(address sender, address remoteUSDCBridge);
+    error ZeroAddressNotAllowed(address addr);
+    error RemoteUSDCBridgeNotSet();
+    error SenderBalanceTooLow(uint256 amount, uint256 balance);
+    error SameMessageServiceAddr(address messageService);
+    error RemoteUSDCBridgeAlreadySet(address remoteUSDCBridge);
+
+    /**
+     * @dev Sends the sender's USDC from L1 to L2, locks the USDC sent
+     * in this contract and sends a message to the message bridge
+     * contract to mint the equivalent USDC on L2
+     * @param amount The amount of USDC to send
+     */
+    function deposit(uint256 amount) external;
+
+    /**
+     * @dev Sends the sender's USDC from L1 to the recipient on L2, locks the USDC sent
+     * in this contract and sends a message to the message bridge
+     * contract to mint the equivalent USDC on L2
+     * @param amount The amount of USDC to send
+     * @param to The recipient's address to receive the funds
+     */
+    function depositTo(uint256 amount, address to) external payable;
+
+    /**
+     * @dev This function is called by the message bridge when transferring USDC from L2 to L1
+     * It burns the USDC on L2 and unlocks the equivalent USDC from this contract to the recipient
+     * @param recipient The recipient to receive the USDC on L1
+     * @param amount The amount of USDC to receive
+     */
+    function receiveFromOtherLayer(address recipient, uint256 amount) external;
+}
+
+// solhint-disable-next-line contract-name-camelcase
+contract Linea_Adapter is AdapterInterface {
+    using SafeERC20 for IERC20;
+
+    WETH9Interface public immutable l1Weth;
+    IERC20 public immutable l1Usdc;
+
+    IMessageService public immutable l1MessageService;
+    ITokenBridge public immutable l1TokenBridge;
+    IUSDCBridge public immutable l1UsdcBridge;
+
+    /**
+     * @notice Constructs new Adapter.
+     * @param _l1Weth WETH address on L1.
+     * @param _l1Usdc USDC address on L1.
+     * @param _l1MessageService Canonical message service contract on L1.
+     * @param _l1TokenBridge Canonical token bridge contract on L1.
+     * @param _l1UsdcBridge L1 USDC Bridge to ConsenSys's L2 Linea.
+     */
+    constructor(
+        WETH9Interface _l1Weth,
+        IERC20 _l1Usdc,
+        IMessageService _l1MessageService,
+        ITokenBridge _l1TokenBridge,
+        IUSDCBridge _l1UsdcBridge
+    ) {
+        l1Weth = _l1Weth;
+        l1Usdc = _l1Usdc;
+        l1MessageService = _l1MessageService;
+        l1TokenBridge = _l1TokenBridge;
+        l1UsdcBridge = _l1UsdcBridge;
+    }
+
+    /**
+     * @notice Send cross-chain message to target on Linea.
+     * @param target Contract on Linea that will receive message.
+     * @param message Data to send to target.
+     */
+    function relayMessage(address target, bytes calldata message) external payable override {
+        l1MessageService.sendMessage{ value: msg.value }(target, 0, message);
```

Comment:
Typically, we don't require the dataworker to send in the ETH on every execution. This has a few problems:
1. Hard to compute this offchain.
2. Can't use the in-contract multicall function because it doesn't allow msg.value and all the value would be used up in the first call.

The way we typically do this is shown in the arbitrum adapter: https://github.com/across-protocol/contracts-v2/blob/866d09fcae4c3257c9759fd150c2b8c6f10034d7/contracts/chain-adapters/Arbitrum_Adapter.sol#L187-L190. We pre-load the dataworker with ETH regularly and compute the value needed in relayMessage in the contract.

Note: remember this code is run via delegatecall, so it's as if it's running within the HubPool contract. msg.value, msg.sender, etc are all the same as they are when code is executing within the HubPool.

Reply:
Got it! Yes, because we need to finalize the relayed messages on Linea ourselves anyways, we are not required to send any value. So in our case, `msg.value` would be `0`. Should I still remove the forwarding of msg.value?

Reply:
Yes I think we should remove any msg.value and just use any ETH held in the HubPool account if we need any. If we don't need any, then just set to 0?

Reply:
Agreed. Remember that these adapters are easy to replace if the requirements change.

Reply:
Bump on this, we should drop the msg.value here